fix: show forwarded message in logviewer.

lorenzo132 opened the pull request:
Forwarded message now will show correctly in logviewer.
<img width="2755" height="678" alt="image" src="https://github.com/user-attachments/assets/e187406e-d869-4b56-a0c0-da4f08a70dce" />

lorenzo132 committed:
fix: show forwarded message in logviewer.

Signed-off-by: lorenzo132 <50767078+lorenzo132@users.noreply.github.com>

lorenzo132 committed:
fix: show forwarded message in logviewer.

Signed-off-by: lorenzo132 <50767078+lorenzo132@users.noreply.github.com>

lorenzo132 committed:
fix: show forwarded message in logviewer.

Signed-off-by: lorenzo132 <50767078+lorenzo132@users.noreply.github.com>

lorenzo132 committed:
black formatting

Signed-off-by: lorenzo132 <50767078+lorenzo132@users.noreply.github.com>

StephenDaDev reviewed (dismissed):
LGTM.

SoAJeff commented on core/utils.py:
```diff
@@ -658,39 +658,49 @@ def extract_forwarded_content(message) -> typing.Optional[str]:
+                    formatted_part += f"[Attachments: {attachment_info}]\n"
+
+                # Add source link to the container message since snapshot doesn't have its own public link
+                formatted_part += f"\n**Source:** {message.jump_url}"
```
This always returns a link to the forwarded message instance that is in the DM with the bot, rather than to the original message that was forwarded - which won't be accessible to the moderator reviewing the content in the logviewer.

The "Context" line when a forwarded message is sent in a Modmail ticket channel (which appears when the forwarded message location is one that the bot is able to fetch) would be a better solution here - perhaps something like what is done in [thread.py](https://github.com/modmail-dev/Modmail/blob/6e75c448f24390bfd27990fbd33729e9295099cf/core/thread.py#L1855)

lorenzo132 replied:
That would be a nice solution yea. If you are able to PR this that would be great, otherwise i will look into this either this week or next week.

SoAJeff reviewed (commented):
Appreciate the quick turn-around on getting a fix for this.  As a mod in a server that receives a lot of forwarded messages in Modmail threads to us, the recent additions in supporting forwarded messages both on the bot side and the logviewer side have been very helpful to us.

Just one minor correction I saw to improve usability for mods reviewing Modmail logs after the fact.

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/forwarded-msg-logviewer

lorenzo132 committed:
remove jump_url

I believe it is a safer aproach to remove the jumpurl from logviewer, since it shows in the threadchannel already.

lorenzo132 committed:
feat: adds attachment support to forwarded content.

lorenzo132 committed:
black formatting

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/forwarded-msg-logviewer

sebkuip commented on core/utils.py:
```diff
@@ -640,6 +641,37 @@ def __init__(self):
+    List[Tuple[str, str]]
+        List of (url, filename) tuples for attachments.
+    """
+    import discord
```
Why do we have an import in the middle of a function? Move this to the top of the file.

sebkuip commented on core/utils.py:
```diff
@@ -640,6 +641,37 @@ def __init__(self):
+
+    attachments = []
+    try:
+        if getattr(message, "message_snapshots", None):
```
Using hasattr() might be a cleaner solution for these checks. Internally python uses getattr() to make it work, but it's better for readability.
https://docs.python.org/3/library/functions.html#hasattr

Please also fix the rest of these instances (I won't be commenting on them all.)

lorenzo132 replied:
Resolved.

sebkuip commented on core/utils.py:
```diff
@@ -640,6 +641,37 @@ def __init__(self):
+                        url = getattr(a, "url", None)
+                        filename = getattr(a, "filename", "Unknown")
+                        if url:
+                            attachments.append((url.split("?")[0], filename))
```
Not sure if it's because of this PR, but forwarded attachments show up incorrectly formatted in the logviewer. They show up as `[image](link)` and don't properly embed. An image sent directly does show up properly.

<img width="787" height="121" alt="Image" src="https://github.com/user-attachments/assets/f4603752-c045-4b18-a81a-ef71d33f8c87" />

lorenzo132 replied:
I have been able to partialy resolve this request. Images would show as following for some reason.
<img width="112" height="69" alt="image" src="https://github.com/user-attachments/assets/5e62c781-6449-4868-bb7c-9daea76c0f1b" />

lorenzo132 committed:
use: hassattr(), remove dedunant package

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/forwarded-msg-logviewer

sebkuip commented on core/clients.py:
```diff
@@ -661,9 +661,18 @@ async def append_log(
         channel_id: str = "",
         type_: str = "thread_message",
     ) -> dict:
+        from core.utils import extract_forwarded_content
```
Import needs to be moved to the top of the file. I don't really see any real reason to have this inside the method here as it's not an expensive import that would slow down startup time significantly.

sebkuip reviewed (changes requested):
I've tested the message forwarding locally and it seems to work just fine. I find one issue with the code regarding the placement of an import. For the rest it looks all good. If you can get this one change fixed up it should be all good to merge.

lorenzo132 committed:
rebase import.

lorenzo132 committed:
fix: linting

sebkuip reviewed (approved):
LGTM now. All thing pointed out have been resolved.

StephenDaDev reviewed (approved):
Looks good, tested working.

sebkuip committed:
Merge branch 'development' into users/lorenzo/forwarded-msg-logviewer
